### Analyzing Used Car Listings on eBay Kleinanzeigen

We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

##### The data dictionary provided with data is as follows:

1) dateCrawled - When this ad was first crawled. All field-values are taken from this date.

2) name - Name of the car.

3) seller - Whether the seller is private or a dealer.

4) offerType - The type of listing

5) price - The price on the ad to sell the car. 

6) abtest - Whether the listing is included in an A/B test.

7) vehicleType - The vehicle Type.

8) yearOfRegistration - The year in which which year the car was first registered.

9) gearbox - The transmission type.

10) powerPS - The power of the car in PS.

11) model - The car model name.

12) kilometer - How many kilometers the car has driven.

13) monthOfRegistration - The month in which which year the car was first registered.

14) fuelType - What type of fuel the car uses.

15) brand - The brand of the car.

16) notRepairedDamage - If the car has a damage which is not yet repaired.

17) dateCreated - The date on which the eBay listing was created.

18) nrOfPictures - The number of pictures in the ad.

19) postalCode - The postal code for the location of the vehicle.

20) lastSeenOnline - When the crawler saw this ad last online.


#### MISSION/TASK
The aim of this project is to clean the data and analyze the included used car listings and also become familiar with some of the unique benefits jupyter notebook provides for pandas.


In [1]:
import pandas as pd
import numpy as np

In [2]:
autos=pd.read_csv('autos.csv',encoding='Latin-1')
autos.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

## OBSERVATIONS

1) Our dataset has 20 columns , 15 of these columns have data stored as a string type and the remaining 5 have data stored as an int type.

2) The columns gearbox,vehicleType,,model,fuelType,notRepairedDamage have null values.this is because the expected value of non null value for each column was to be 50000.No columns has more than ~20% null values

3) Expected the price column to be an int64/float type but it is an object type

4) The column names use camelcase instead of Python's preferred snakecase , which means we can't just replace spaces with underscores.<< https://en.wikipedia.org/wiki/Snake_case
<< https://en.wikipedia.org/wiki/Camel_case

5) Both dateCrawled and lastSeen columns are in object (string) format and not in datetime. 
 


In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos


dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

## CLEANING AND REPLACING COLUMN NAMES

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
col_name_copy=autos.columns.copy()

#### Changes made
Change the columns from camelcase to snakecase.

Change a few wordings to more accurately describe the columns.

In [7]:
def col_name_edited(col_name):
    col_name=col_name.replace('yearOfRegistration','registration_year')
    col_name=col_name.replace('monthOfRegistration','registration_month')
    col_name=col_name.replace('notRepairedDamage','unrepaired_damage')
    col_name=col_name.replace('dateCreated','ad_created')
    col_name=col_name.replace('nrOfPictures','num_photos')
    col_name=col_name.replace("dateCrawled", "date_crawled")
    col_name=col_name.replace("offerType", "offer_type")
    col_name=col_name.replace("vehicleType", "vehicle_type")
    col_name=col_name.replace("powerPS", "power_ps")
    col_name=col_name.replace("fuelType", "fuel_type")
    col_name=col_name.replace("postalCode", "postal_code")
    col_name=col_name.replace("lastSeen", "last_seen")
    col_name=col_name.lower()
    return col_name
    
new_col_name=[]
for name in col_name_copy:
    edited=col_name_edited(name)
    new_col_name.append(edited)
print(new_col_name)
    

['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code', 'last_seen']


In [8]:
autos.columns=new_col_name

In [9]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## EXPLORING DATA

###### Descriptive statstitics for all columns

In [11]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

#### Analysis to determine which column to clean /drop

1) Any columns that have mostly one value that are candidates to be dropped

In [12]:
#FROM THE DATAFRAME.describe() function it can be observed that 
# 1) the seller column has only 2 unique values,of which privat is the unique value with the most occurance of 49999 out of 50000(the total number of entries). hence the other unique value occurs only once.

# 2) the offer_type column has only 2 unique values,of which Angebot is the unique value with the most occurance of 49999 out of 50000(the total number of entries). hence the other unique value occurs only once.

seller_column=autos['seller'].value_counts()
print(seller_column)

print('\n')
offer_type_column=autos['offer_type'].value_counts()
print(offer_type_column)

# nb.Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.


privat        49999
gewerblich        1
Name: seller, dtype: int64


Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64


2) Any columns that need more investigation.

The num_photos column

In [13]:
# num_photos is an int type object as such the mean,std,min,percentiles and max was calculated as was the case for the other int type columns( registration_year,power_ps,registration_month,postal_codes)
# but it returned 0.0 despite the fact that it had no unique value() but had 50000 entries/non null values

num_photos_col=autos['num_photos'].value_counts()
print(num_photos_col)

#the num_photos column has 0 for every row.

0    50000
Name: num_photos, dtype: int64


##### Dropping columns 
As we can see, the columns seller, offer_type and num_photos have mostly one value: privat, Angebot and 0, respectively. So, we'll drop them.

In [14]:
autos=autos.drop(["num_photos", "seller", "offer_type"],axis=1)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

3) Any examples of numeric data stored as text that needs to be cleaned

That is the price and odometer columns have int values which have been stored as string or object type values due to the $/, for the price and km/, for the odometer 
the price column is strange too as it has most of its values to be 0.

In [15]:
#Expection of columns to be change to int type

cols=autos[['odometer','price']]
print(cols)

        odometer    price
0      150,000km   $5,000
1      150,000km   $8,500
2       70,000km   $8,990
3       70,000km   $4,350
4      150,000km   $1,350
...          ...      ...
49995  100,000km  $24,900
49996  150,000km   $1,980
49997    5,000km  $13,200
49998   40,000km  $22,900
49999  150,000km   $1,250

[50000 rows x 2 columns]


In [16]:
#removing non int characters for both odometer and price columns

autos['price']=autos['price'].str.replace(',','').str.replace('$','').astype(int)
autos['odometer']=autos['odometer'].str.replace(',','').str.replace('km','').astype(int)

In [17]:
autos[['odometer','price']].dtypes

odometer    int32
price       int32
dtype: object

In [18]:
autos = autos.rename({"odometer": "odometer_km"}, axis=1)
autos = autos.rename({"price": "price_$"}, axis=1)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price_$   abtest vehicle_type  registration_year    gearbox  power_ps  \
0     5000  control          bus               2004    manuell       158   
1     8500  control    limousine               1997  automatik       286   
2     8990     test    limousine               2009    manuell       102   
3     4350  control   kleinwagen               2007  automatik        71   
4     1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

#### Other Comments

There are columns that seem to have some weird values:

1) price: the most frequent value is $0, but (we assume that) no one will put a car on sale for nothing...

2) registration_year: the lowest and highest registration year are 1000 and 9999, which clearly there are innacuracies on data.

3) power_ps: again, the lowest and highest values are impossible to happen, 0 and 17700.

4) registration_month: the minimum value is 0, which should be 1 (january)

5)the columns with null values indicated early on must also be looked at.The columns gearbox,vehicleType,,model,fuelType,notRepairedDamage have null values.this is because the expected value of non null value for each column was to be 50000.No columns has more than ~20% null values

## ANALYZING COLUMNS

###### ODOMETER COLUMN

In [19]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos['odometer_km'].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [21]:
autos['odometer_km'].unique().shape
#13 unique values

(13,)

In [22]:
autos['odometer_km'].value_counts(dropna=False)
#it has no null values

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

##### PRICE COLUMNS

In [23]:
autos['price_$'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

In [24]:
autos['price_$'].unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

In [25]:
autos['price_$'].unique().shape

(2357,)

Note the unique values in the price columns is endless as such value_counts() would be endlessly long.to reduce its lenght ,head() would be used

In [26]:
autos['price_$'].value_counts(dropna=False).head(20)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price_$, dtype: int64

Again, the prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.

There are 1,421 cars listed with $0 price - given that this is only 2% of the of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [27]:

# viewing the lowest and highest unique values with their counts
autos['price_$'].value_counts(dropna=False).sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_$, dtype: int64

In [28]:

# viewing the lowest and highest unique values with their counts
autos['price_$'].value_counts(dropna=False).sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_$, dtype: int64

#### OUTLIERS

There are a lot of entries with a price of 0. It seems unlikely that cars were actually available for nothing, and that these simply didn't advertise the price acurately.

In addition, there are a few cars that are priced unbelievably high. It seems unlikely that a car was priced at 99,999,999 dollars. Research into luxury car auctions, show that rare, collectible cars can sell for millions of dollars. The most expensive car sold was 48 million dollars. However, these are taking place at premium auction sites such as Sotheby's. It seems unlikely that a site like eBay would be selling a car of the same caliber. 

Given that eBay is an auction site, there could legitimately be items where the opening bid is 1 dollar. We will keep the 1 dollar items, but remove anything above 350,000 dollars since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [29]:
autos=autos[autos['price_$'].between(1,351000)]
autos['price_$'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64

In [30]:
autos.shape

(48565, 17)

#### DATE COLUMNS

There are 5 columns that represent date values. Differentiating them, we have:

date_crawled: added by the crawler

last_seen: added by the crawler

ad_created: from the website

registration_month: from the website

registration_year: from the website


The date_crawled, last_seen and ad_created columns are represented as string values by pandas. We need to convert them to numerical representation in order to understand it quantitatively. The other two columns are represented as numeric values so, we can use methods like Series.describe() to understand the distribution without any extra data processing.


In [31]:
##### non numeric columns

In [32]:
autos[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [33]:
#analysing date_crawled column
#value_counts as percentages

autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False)

2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [34]:
autos['date_crawled'].unique().shape

(46882,)

In [35]:
# rank by date in ascending order (earliest to latest)

autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [36]:
#analysing last_seen column
#value_counts as percentages

autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False)

2016-04-06    0.221806
2016-04-07    0.131947
2016-04-05    0.124761
2016-03-17    0.028086
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024771
2016-04-04    0.024483
2016-03-12    0.023783
2016-03-31    0.023783
2016-04-01    0.022794
2016-03-29    0.022341
2016-03-22    0.021373
2016-03-28    0.020859
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018532
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012602
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

In [37]:
# rank by date in ascending order (earliest to latest)

autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The ads were last seen on the same dates as the data was scraped, March, 5, 2016 to April, 7, 2016. Distribution is relatively even.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [38]:
#analysing ad_created column
#value_counts as percentages

autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False)

2016-04-03    0.038855
2016-03-20    0.037949
2016-03-21    0.037579
2016-04-04    0.036858
2016-03-12    0.036755
                ...   
2016-02-09    0.000021
2015-12-05    0.000021
2016-02-17    0.000021
2016-02-01    0.000021
2015-11-10    0.000021
Name: ad_created, Length: 76, dtype: float64

In [39]:
# rank by date in ascending order (earliest to latest)

autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

It seems that before March-2016 there were few ads, less than 10 publications, and as soon this month began it started to appear more and more ads.

The ads were created between June 2015 and April 2016. The majority of the ads were created in 2016, predominatly March and the first few days of April.

##### the numeric date columns

The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.



In [40]:
# registration_year

autos['registration_year'].describe()



count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

###### SOLUTION TO CLEANING REGISTRATION YEAR DATA

Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column

In [41]:

(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

Given that this is less than 4% of our data, we will remove these rows.

In [42]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [43]:
autos['registration_year'].value_counts(normalize=True).sort_index()

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: registration_year, Length: 78, dtype: float64

The marjority of the cars were registered in the mid 1990's through 2016.

## EXPLORING PRICE BY BRAND

In [44]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

#### TOP 20

In [45]:
a=autos['brand'].value_counts(normalize=True).head(20)
a

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
Name: brand, dtype: float64

In [46]:
b20=a.index

In [47]:
brand20_mean_prices={}

for brand_name in b20:
    brand_bool=autos['brand']==brand_name
    mean_price=autos.loc[brand_bool,'price_$'].mean()
    brand20_mean_prices[brand_name]=int(mean_price)
print(brand20_mean_prices)

{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749, 'renault': 2474, 'peugeot': 3094, 'fiat': 2813, 'seat': 4397, 'skoda': 6368, 'nissan': 4743, 'mazda': 4112, 'smart': 3580, 'citroen': 3779, 'toyota': 5167, 'hyundai': 5365, 'sonstige_autos': 12338, 'volvo': 4946, 'mini': 10613}


###### BRANDS WHICH OCCUPIES MORE THEN 50% OF THE TOTAL LISTING

There are 40 different car brands, whereas four out of the top five brands are from German manufacturers, which represents more than 50% of the total announced in the german eBay.

Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [48]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [49]:
brand_mean_prices={}

for brand_name in common_brands:
    brand_bool=autos['brand']==brand_name
    mean_price=autos.loc[brand_bool,'price_$'].mean()
    brand_mean_prices[brand_name]=int(mean_price)
print(brand_mean_prices)

{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749}


Of the top 5 brands, there is a distinct price gap:

Audi, BMW and Mercedes Benz are more expensive
Ford and Opel are less expensive
Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

###### comapring brand,price and mileage for brands representing more than 5% of total listings.

#### method 1

In [50]:
brand_price_mile={}

for brand_name in common_brands:
    brand_bool=autos['brand']==brand_name
    mean_price=autos.loc[brand_bool,'price_$'].mean()
    mean_mileage=autos.loc[brand_bool,'odometer_km'].mean()
    brand_price_mile[brand_name]=(int(mean_price),int(mean_mileage))
print(brand_price_mile)

{'volkswagen': (5402, 128707), 'bmw': (8332, 132572), 'opel': (2975, 129310), 'mercedes_benz': (8628, 130788), 'audi': (9336, 129157), 'ford': (3749, 124266)}


In [51]:

df = pd.DataFrame(brand_price_mile,index=[['mean_price','mileage']])
df

volkswagen     bmw    opel  mercedes_benz    audi    ford
mean_price        5402    8332    2975           8628    9336    3749
mileage         128707  132572  129310         130788  129157  124266

#### method 2


In [52]:
brand_mean_prices={}

for brand_name in common_brands:
    brand_bool=autos['brand']==brand_name
    mean_price=autos.loc[brand_bool,'price_$'].mean()
    brand_mean_prices[brand_name]=int(mean_price)
print(brand_mean_prices)

{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749}


In [53]:
bmp_series=pd.Series(brand_mean_prices)
bmp_series

volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
dtype: int64

In [54]:
brand_mean_mileage={}

for brand_name in common_brands:
    brand_bool=autos['brand']==brand_name
    mean_mileage=autos.loc[brand_bool,'odometer_km'].mean()
    brand_mean_mileage[brand_name]=int(mean_mileage)
print(brand_mean_mileage)

{'volkswagen': 128707, 'bmw': 132572, 'opel': 129310, 'mercedes_benz': 130788, 'audi': 129157, 'ford': 124266}


In [55]:
bmm_series=pd.Series(brand_mean_mileage)
bmm_series

volkswagen       128707
bmw              132572
opel             129310
mercedes_benz    130788
audi             129157
ford             124266
dtype: int64

In [56]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [57]:
df['mileage']=bmm_series
df

mean_price  mileage
volkswagen           5402   128707
bmw                  8332   132572
opel                 2975   129310
mercedes_benz        8628   130788
audi                 9336   129157
ford                 3749   124266

In [58]:
df.sort_values('mean_price',ascending=False)

mean_price  mileage
audi                 9336   129157
mercedes_benz        8628   130788
bmw                  8332   132572
volkswagen           5402   128707
ford                 3749   124266
opel                 2975   129310

The range of car mileages does not vary as much as the prices do by brand. there is no relationship between the price and milage stratified by the brands

## EXTRAS

#### question 1
Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.

the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively

In [59]:
autos['date_crawled'].str.split().str[0].str.replace('-','').astype(int)

0        20160326
1        20160404
2        20160326
3        20160312
4        20160401
           ...   
49995    20160327
49996    20160328
49997    20160402
49998    20160308
49999    20160314
Name: date_crawled, Length: 46681, dtype: int32

In [60]:
autos['last_seen'].str[:10].str.replace('-','').astype(int)

0        20160406
1        20160406
2        20160406
3        20160315
4        20160401
           ...   
49995    20160401
49996    20160402
49997    20160404
49998    20160405
49999    20160406
Name: last_seen, Length: 46681, dtype: int32

In [61]:
autos['ad_created'].str.split().str[0].str.replace('-','').astype(int)

0        20160326
1        20160404
2        20160326
3        20160312
4        20160401
           ...   
49995    20160327
49996    20160328
49997    20160402
49998    20160308
49999    20160313
Name: ad_created, Length: 46681, dtype: int32

##### question 2 
See if there are particular keywords in the name column that you can extract as new columns

In [92]:
new=autos[['name','brand','model']].head(30)
new

name          brand     model
0                    Peugeot_807_160_NAVTECH_ON_BOARD        peugeot    andere
1          BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik            bmw       7er
2                          Volkswagen_Golf_1.6_United     volkswagen      golf
3   Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...          smart    fortwo
4   Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...           ford     focus
5   Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...       chrysler   voyager
6   VW_Golf_III_GT_Special_Electronic_Green_Metall...     volkswagen      golf
7                                Golf_IV_1.9_TDI_90PS     volkswagen      golf
8                                          Seat_Arosa           seat     arosa
9           Renault_Megane_Scenic_1.6e_RT_Klimaanlage        renault    megane
11                         Mercedes_A140_Motorschaden  mercedes_benz       NaN
12  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...          smart    fortwo
13                                 Audi_A3_1.6_tuning           audi        a3
14  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...        renault      clio
15        Corvette_C3_Coupe_T_Top_Crossfire_Injection          other       NaN
16                                Opel_Vectra_B_Kombi           opel    vectra
17                          Volkswagen_Scirocco_2_G60     volkswagen  scirocco
18               Verkaufen_mein_bmw_e36_320_i_touring            bmw       3er
19        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018          mazda    andere
20  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...           audi        a4
21                      Porsche_911_Carrera_4S_Cabrio        porsche       911
22                               MINI_Cooper_S_Cabrio           mini    cooper
23   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima        peugeot       NaN
24                         BMW_535i_xDrive_Sport_Aut.            bmw       5er
25  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung           ford       NaN
26                                Volkswagen_Polo_Fox     volkswagen      polo
28                                      MINI_Cooper_D           mini    cooper
29  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...  mercedes_benz  e_klasse
30                         Peugeot_206_Unfallfahrzeug        peugeot   2_reihe
31                          Noch_gut_erhaltenen_C_320  mercedes_benz  c_klasse

In [63]:
split=new['name'].str.split('_').str[0]
split

0        Peugeot
1            BMW
2     Volkswagen
3          Smart
4           Ford
5       Chrysler
6             VW
7           Golf
8           Seat
9        Renault
11      Mercedes
12         Smart
13          Audi
14       Renault
15      Corvette
16          Opel
17    Volkswagen
18     Verkaufen
19         mazda
20          Audi
21       Porsche
22          MINI
23       Peugeot
24           BMW
25          Ford
26    Volkswagen
28          MINI
29      Mercedes
30       Peugeot
31          Noch
Name: name, dtype: object

In [64]:
new['split_first_name']=split

In [65]:
new

name           brand  \
0                    Peugeot_807_160_NAVTECH_ON_BOARD         peugeot   
1          BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik             bmw   
2                          Volkswagen_Golf_1.6_United      volkswagen   
3   Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...           smart   
4   Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...            ford   
5   Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...        chrysler   
6   VW_Golf_III_GT_Special_Electronic_Green_Metall...      volkswagen   
7                                Golf_IV_1.9_TDI_90PS      volkswagen   
8                                          Seat_Arosa            seat   
9           Renault_Megane_Scenic_1.6e_RT_Klimaanlage         renault   
11                         Mercedes_A140_Motorschaden   mercedes_benz   
12  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...           smart   
13                                 Audi_A3_1.6_tuning            audi   
14  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...         renault   
15        Corvette_C3_Coupe_T_Top_Crossfire_Injection  sonstige_autos   
16                                Opel_Vectra_B_Kombi            opel   
17                          Volkswagen_Scirocco_2_G60      volkswagen   
18               Verkaufen_mein_bmw_e36_320_i_touring             bmw   
19        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018           mazda   
20  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...            audi   
21                      Porsche_911_Carrera_4S_Cabrio         porsche   
22                               MINI_Cooper_S_Cabrio            mini   
23   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima         peugeot   
24                         BMW_535i_xDrive_Sport_Aut.             bmw   
25  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung            ford   
26                                Volkswagen_Polo_Fox      volkswagen   
28                                      MINI_Cooper_D            mini   
29  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   mercedes_benz   
30                         Peugeot_206_Unfallfahrzeug         peugeot   
31                          Noch_gut_erhaltenen_C_320   mercedes_benz   

       model split_first_name  
0     andere          Peugeot  
1        7er              BMW  
2       golf       Volkswagen  
3     fortwo            Smart  
4      focus             Ford  
5    voyager         Chrysler  
6       golf               VW  
7       golf             Golf  
8      arosa             Seat  
9     megane          Renault  
11       NaN         Mercedes  
12    fortwo            Smart  
13        a3             Audi  
14      clio          Renault  
15       NaN         Corvette  
16    vectra             Opel  
17  scirocco       Volkswagen  
18       3er        Verkaufen  
19    andere            mazda  
20        a4             Audi  
21       911          Porsche  
22    cooper             MINI  
23       NaN          Peugeot  
24       5er              BMW  
25       NaN             Ford  
26      polo       Volkswagen  
28    cooper             MINI  
29  e_klasse         Mercedes  
30   2_reihe          Peugeot  
31  c_klasse             Noch

#### question 3
convert date columns to datetime 

In [66]:
autos_copy=autos.copy()
autos_copy[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We can notice that a date is in a format YYYY-MM-DD HH-MM-SS. We are going to use pd.to_datetime to change astype to datetime and we will change format to YYYY-MM-DD.

##### data_crawled

In [67]:
autos_copy['date_crawled'] = pd.to_datetime(autos_copy['date_crawled'], format="%Y-%m-%d %H:%M:%S") # change dtype to datetime
date_crawled = autos_copy['date_crawled'].dt.strftime('%Y-%m-%d') # change format to YYYY-MM-DD
date_crawled

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 46681, dtype: object

ad_created

In [68]:
autos_copy['ad_created'] = pd.to_datetime(autos_copy['ad_created'], format="%Y-%m-%d %H:%M:%S") # change dtype to datetime
ad_created= autos_copy['ad_created'].dt.strftime('%Y-%m') # change format to YYYY-MM-DD
ad_created

0        2016-03
1        2016-04
2        2016-03
3        2016-03
4        2016-04
          ...   
49995    2016-03
49996    2016-03
49997    2016-04
49998    2016-03
49999    2016-03
Name: ad_created, Length: 46681, dtype: object

last_seen


In [69]:
autos_copy['last_seen'] = pd.to_datetime(autos_copy['last_seen'], format="%Y-%m-%d %H:%M:%S")
last_seen = autos_copy['last_seen'].dt.strftime("%Y-%m-%d")
last_seen

0        2016-04-06
1        2016-04-06
2        2016-04-06
3        2016-03-15
4        2016-04-01
            ...    
49995    2016-04-01
49996    2016-04-02
49997    2016-04-04
49998    2016-04-05
49999    2016-04-06
Name: last_seen, Length: 46681, dtype: object

In [70]:
autos_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46681 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   date_crawled        46681 non-null  datetime64[ns]
 1   name                46681 non-null  object        
 2   price_$             46681 non-null  int32         
 3   abtest              46681 non-null  object        
 4   vehicle_type        43977 non-null  object        
 5   registration_year   46681 non-null  int64         
 6   gearbox             44571 non-null  object        
 7   power_ps            46681 non-null  int64         
 8   model               44488 non-null  object        
 9   odometer_km         46681 non-null  int32         
 10  registration_month  46681 non-null  int64         
 11  fuel_type           43363 non-null  object        
 12  brand               46681 non-null  object        
 13  unrepaired_damage   38374 non-null  object    

#### question 4

Identify categorical data that uses german words, translate them and map the values to their english counterparts

We are going to translate data from this columns: vehicle_type, gearbox, model, fuel_type, brand and unrepaired_damage from german to english. First we will check unique values for every column, translate them and than map the values to their english counterparts.

In [71]:
autos[['vehicle_type', 'gearbox', 'model','fuel_type', 'brand','unrepaired_damage']]

vehicle_type    gearbox   model fuel_type       brand unrepaired_damage
0              bus    manuell  andere       lpg     peugeot              nein
1        limousine  automatik     7er    benzin         bmw              nein
2        limousine    manuell    golf    benzin  volkswagen              nein
3       kleinwagen  automatik  fortwo    benzin       smart              nein
4            kombi    manuell   focus    benzin        ford              nein
...            ...        ...     ...       ...         ...               ...
49995    limousine  automatik      q5    diesel        audi              nein
49996       cabrio    manuell   astra    benzin        opel              nein
49997       cabrio  automatik     500    benzin        fiat              nein
49998        kombi    manuell      a3    diesel        audi              nein
49999    limousine    manuell  vectra    benzin        opel              nein

[46681 rows x 6 columns]

vehicle type

In [72]:
autos['vehicle_type'].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [73]:
# create dictionary with translations
vt_dict = {
    'limousine': 'limousine',
    'kleinwagen': 'small_car',
    'kombi': 'combi',
    'bus': 'bus',
    'cabrio': 'convertible',
    'coupe': 'coupe',
    'suv': 'suv',
    'andere': 'other'
}

In [74]:
autos['vehicle_type']=autos['vehicle_type'].map(vt_dict)

In [75]:
autos['vehicle_type'].unique()

array(['bus', 'limousine', 'small_car', 'combi', nan, 'coupe', 'suv',
       'convertible', 'other'], dtype=object)

gearbox

In [76]:
autos['gearbox'].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [77]:
# create dictionary with translations
gb_dict = {
    'manuell': 'manual',
    'automatik': 'automatic'
}
autos['gearbox'] = autos['gearbox'].map(gb_dict) # map the new values
print(pd.unique(autos['gearbox'])) # check new values

['manual' 'automatic' nan]


fuel_type

In [78]:
autos['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [79]:
# create dictionary with translations
ft_dict = {
    'benzin': 'petrol',
    'diesel': 'diesel',
    'lpg': 'lpg',
    'cng': 'cng',
    'hybrid': 'hybrid',
    'elektro': 'electric',
    'andere': 'other'
}
autos['fuel_type'] = autos['fuel_type'].map(ft_dict) # map the new values
print(pd.unique(autos['fuel_type'])) # check new values

['lpg' 'petrol' 'diesel' nan 'cng' 'hybrid' 'electric' 'other']


unrepaired_damaged

In [80]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In [81]:
# create dictionary with translations
ud_dict = {
    'ja': 'yes',
    'nein': 'no'
}
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(ud_dict) # map the new values
print(pd.unique(autos['unrepaired_damage'])) # check new values

['no' nan 'yes']


model

In [82]:
autos['model'].unique()

array(['andere', '7er', 'golf', 'fortwo', 'focus', 'voyager', 'arosa',
       'megane', nan, 'a3', 'clio', 'vectra', 'scirocco', '3er', 'a4',
       '911', 'cooper', '5er', 'polo', 'e_klasse', '2_reihe', 'c_klasse',
       'corsa', 'mondeo', 'altea', 'a1', 'twingo', 'a_klasse', 'cl',
       '3_reihe', 's_klasse', 'sandero', 'passat', 'primera', 'wrangler',
       'a6', 'transporter', 'astra', 'v40', 'ibiza', 'micra', '1er',
       'yaris', 'colt', '6_reihe', '5_reihe', 'corolla', 'ka', 'tigra',
       'punto', 'vito', 'cordoba', 'galaxy', '100', 'octavia', 'm_klasse',
       'lupo', 'fiesta', 'superb', 'meriva', 'c_max', 'laguna', 'touran',
       '1_reihe', 'm_reihe', 'touareg', 'seicento', 'avensis', 'vivaro',
       'x_reihe', 'ducato', 'carnival', 'boxster', 'signum', 'sharan',
       'zafira', 'rav', 'a5', 'beetle', 'c_reihe', 'phaeton', 'i_reihe',
       'sl', 'insignia', 'up', 'civic', '80', 'mx_reihe', 'omega',
       'sorento', 'z_reihe', 'berlingo', 'clk', 's_max', 'kalos',
 

Since we have alot unique values and we want translate only one type andere, we will use str.replace()

In [83]:
autos['brand'] = autos['brand'].str.replace('sonstige_autos','other')
print(pd.unique(autos['brand']))

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'other' 'opel' 'mazda' 'porsche' 'mini' 'toyota'
 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar' 'fiat'
 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda' 'daewoo'
 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover' 'daihatsu'
 'lancia']


In [84]:
autos_copy.columns

Index(['date_crawled', 'name', 'price_$', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [85]:
autos_copy.columns.tolist()

['date_crawled',
 'name',
 'price_$',
 'abtest',
 'vehicle_type',
 'registration_year',
 'gearbox',
 'power_ps',
 'model',
 'odometer_km',
 'registration_month',
 'fuel_type',
 'brand',
 'unrepaired_damage',
 'ad_created',
 'postal_code',
 'last_seen']

#### Find the most common brand/model combinations

In [86]:
k=autos[['brand','model']]

In [87]:
autos[['brand','model']].head(20)

brand     model
0         peugeot    andere
1             bmw       7er
2      volkswagen      golf
3           smart    fortwo
4            ford     focus
5        chrysler   voyager
6      volkswagen      golf
7      volkswagen      golf
8            seat     arosa
9         renault    megane
11  mercedes_benz       NaN
12          smart    fortwo
13           audi        a3
14        renault      clio
15          other       NaN
16           opel    vectra
17     volkswagen  scirocco
18            bmw       3er
19          mazda    andere
20           audi        a4

In [88]:
autos['brand'].describe()



count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object

In [89]:
#40 unique brands
p=autos['brand'].unique()
p

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'other', 'opel',
       'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan', 'jeep',
       'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda', 'subaru',
       'kia', 'citroen', 'chevrolet', 'hyundai', 'honda', 'daewoo',
       'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada', 'rover',
       'daihatsu', 'lancia'], dtype=object)

In [90]:
#unique model for each brand
p=autos['brand'].unique()

brand_model={}
for brand in p:
    r=autos.loc[autos['brand']==brand,'model']
    brand_model[brand]= r.unique()
    
brand_model

{'peugeot': array(['andere', nan, '2_reihe', '3_reihe', '1_reihe', '5_reihe',
        '4_reihe'], dtype=object),
 'bmw': array(['7er', '3er', '5er', nan, '1er', 'm_reihe', 'x_reihe', 'z_reihe',
        'andere', '6er', 'i3'], dtype=object),
 'volkswagen': array(['golf', 'scirocco', 'polo', nan, 'passat', 'transporter', 'lupo',
        'touran', 'touareg', 'sharan', 'beetle', 'phaeton', 'up', 'tiguan',
        'kaefer', 'bora', 'fox', 'andere', 'caddy', 'eos', 'cc', 'jetta',
        'amarok'], dtype=object),
 'smart': array(['fortwo', nan, 'forfour', 'roadster', 'andere'], dtype=object),
 'ford': array(['focus', nan, 'mondeo', 'ka', 'galaxy', 'fiesta', 'c_max', 's_max',
        'transit', 'mustang', 'andere', 'fusion', 'escort', 'kuga',
        'b_max'], dtype=object),
 'chrysler': array(['voyager', 'andere', 'ptcruiser', '300c', nan, 'crossfire',
        'grand'], dtype=object),
 'seat': array(['arosa', 'altea', 'ibiza', 'cordoba', 'mii', 'leon', 'toledo',
        'exeo', nan, 'andere'